### DISCUSSION TITLE
Add `Sendable` conformance to association types

### DISCUSSION THREAD

**Discussion by seanmrich** [groue/GRDB.swift#1492] [Ideas]
In the [associations](https://github.com/groue/GRDB.swift/blob/master/Documentation/AssociationsBasics.md) section of the documentation there is this code on how to define associations:
```swift
extension Author {
  static let books = hasMany(Book.self)
//           ^ Warning: Static property 'books' is not concurrency-safe because it is not either conforming to 'Sendable' or isolated to a global actor; this is an error in Swift 6
  ...
}
```
When using strict concurrency since the implementation of [SE-0412](https://github.com/apple/swift-evolution/blob/main/proposals/0412-strict-concurrency-for-global-variables.md) you get a warning as annotated in the code above. A workaround is to write the declarations as `static var` rather than `static let`, but then you have to fully specify the type and thereby lose the nice ergonomics of the example above. Would it be possible to conform types like `BelongsToAssociation` and `HasManyAssociation` to `Sendable`?

**Comment by groue** [comment]
Thanks for the report, @seanmrich. I haven't installed the latest beta yet.

Strictly speaking, associations are not Sendable, because they can embed non-Sendable values or closures. See for example the family of methods name `-WhenConnected` in [`DerivableRequest`](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/derivablerequest): anything can happen, there. `TableAlias` is not Sendable either, and requires a rework.

Practically speaking, I'm sure that most apps define requests and associations that only embed values and closures that are free of data races.

Still, making them strictly Sendable would introduce a lot of breaking changes, that will require bumping GRDB to GRDB 7.

I'm torn. I wish we could find a step-by-step way forward.

Besides changing `let` to `var`, did you try other workarounds? `nonisolated(unsafe)` (not sexy)? `@preconcurrency import GRDB`? Something else?

**Reply by seanmrich** [reply]
`@preconcurrency` doesn't fix the warning. I don't think I fully understand how this works and it has rarely done anything for libraries I've tried it with. Perhaps because I'm using strict concurrency, or maybe because there is already adoption of structured concurrency features in GRDB? Not sure.

`nonisolated(unsafe)` does remove the warning, but I don't feel comfortable forcing the override. Besides just turning off sendability checking, I'm not sure what else would affect it. My impression is that constructing these associations on demand shouldn't be a significant burden unless I'm missing something?

Looking through the code supporting associations, it looks like many of the model types are trivially `Sendable` now. Perhaps that's a starting point? As you mentioned, `TableAlias` looks like a sticking point and would need to be reconsidered (actor?). As Apple continues to turn the screws on data race safety, `Sendable` is going to spread virally through our code, especially something like GRDB. 

Relatedly, the compiler is also warning about static declarations in GRDB/Core/Database.swift within `ColumnType`. This seems like one of those trivially `Sendable` types that could be converted to remove those warnings.

**Reply by groue** [reply]
Thank you for your checks.

Yes, the way forward is clearly to make Sendable as many GRDB types as possible. I have already explored what it takes, and this is how I know that this can't be done without introducing breaking changes.

Errr... 🤔 Actually I'm not 100% sure if the required changes would introduce hard errors or warnings. Warnings are not a breaking change, but hard errors are (a user upgrades GRDB and their app stops compiling - that's what we want to avoid).

[I just asked on the Swift forums](https://forums.swift.org/t/how-to-introduce-concurrency-annotations-in-a-library-without-introducing-breaking-changes/69957).

**Comment by mattmassicotte** [comment]
Hello!

I am not a GRDB user, and I know basically zero about how it works. However, I am a Swift concurrency "enthusiast" and @groue got me interested because it turns out I do not understand what's going on.

First, why doesn't `@preconcurrency` fix the issue? This particular problem isn't related to the public interfaces of GRDB. The `hasMany` function returns a non-isolated and non-Sendable type. And *that* cannot be stored safely a static variable. Forget about Swift concurrency - what happens if I do this?

```swift
for i in 0..<1000 {
	DispatchQueue.global().async {
		Author.books.someActionOnBooks()
	}
}
```

Is this safe? If so, perhaps the type that `hasMany` returns actually is Sendable! And if not, could it be this is a latent (if unlikely) bug that the compiler is now forcing you to deal with?

**Reply by groue** [reply]
Yes, the goal is indeed to make associations (the type returned by `hasMany`) publicly conform to Sendable. They are not at the present time, because I did not want to break user code by adding Sendable annotations in all the places required for associations to become Sendable. Fortunately, [Holly Borla has lifted all possible doubts](https://forums.swift.org/t/is-adding-a-sendable-requirement-a-breaking-change/69957/7): the language designers WANT us to be able to add Sendable annotations without breaking user code.

@seanmrich, please just wait a little until GRDB 7 ships.

**Reply by seanmrich** [reply]
Great news 👍

**Reply by mattmassicotte** [reply]
That [bug](https://github.com/apple/swift/issues/71544) you found is amazing. Please keep filing those!

My experiences with `@preconcurrency` have been similar: it rarely seems to solve my problems. But, I'm starting to realize this is due to me just not understanding the problems it can/cannot fix. And in fact, the [proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0412-strict-concurrency-for-global-variables.md) that handles this case calls out the behavior very explicitly:

> Importing a module via @preconcurrency import suppresses any potential errors resulting from data isolation checking of imported global variables that lack explicit concurrency annotations. Any use of a @preconcurrency imported concurrency-unsafe global variable will produce a warning at the use site.

So, in Swift 6, `@preconcurrency import` could get this **building** again, but not make it warning-free by design. The correct client-side tool for that is `nonisolated(unsafe)`. I did, however, notice a [bug](https://github.com/apple/swift/issues/71546) that would prevent you from being informed should that `nonisolated(unsafe)` become unnecessary in the future.

Thanks for tolerating my intrusion here! Please carry on!

**Reply by groue** [reply]
Thank you for your great summary :+1:

So, until I ship a version of the library where the involved types are Sendable, I will have to suggest one of those two workarounds:

1. Mark `let` globals as `nonisolated(unsafe)`.
2. Disable strict concurrency checks.

I'm not sure about (2) yet. First, does it remove those warnings? I'll check.

More importantly, is (2) a _good advice_?

Some teams don't really care that much _yet_, and enabled the strict checks because of...  let's say fashion, influent devs on social networks, etc. Disabling the checks may be an acceptable temporary workaround for them.

For the other teams, the ones who care about concurrency checks, it's more complicated:

On one side, Swift 6 will be [merciless](https://hachyderm.io/@miguelcabeca/111913082095624585), so we want to encourage users to perform strict checks early.

On the other side, there are still a lot of false positives that have us introduce useless workarounds in our code. Recent proposals address this problem, e.g. [SE-01414 Region based Isolation](https://github.com/apple/swift-evolution/blob/main/proposals/0414-region-based-isolation.md). Your recent discussions with @hborla on the Swift forums are also very fruitful on this topic.

Because of the false positives, my current position/goal would be:

- Strict checks always on in libraries, their tests, and demo apps. Benefits: no one can open an issue about warnings, _and_ the library author can have a taste of the real life of the library users - empathy is a necessary step towards good apis, good docs, and good support answers. Shape apis so that they don't create unsolvable warnings in user code, and stress in user's mind. Mark the workarounds for what looks like a false positive with a TODO so that it can be removed later.
- Strict checks disabled by default in apps. However, apps should *from time to time* enable strict checks, look for "this will be an error in Swift 6" warnings, study them with a grain of salt and a relaxed mind, and only refactor what can honestly be described as a dirty concurrency mess. This requires guts and skills, though 😅

**Reply by mattmassicotte** [reply]
I think your position for libraries is correct.

For apps, here's how I look at it. If the goal is "no warnings", just turn off the checking. Done! But, if the goal is "correct use of concurrency" I think turning off checking is a huge mistake.

It is just not possible to build a correct mental model about how Swift concurrency works without feedback from the compiler. This will result in building systems that may be fundamentally unfixable. I speak from personal experience. And, because the isolation checking that the compiler does is exhaustive, you might not even realize you are using concurrency (see `static let`).

Further, most of Apple's own APIs are insufficiently or incorrectly annotated. Don't even get me **started** on Obj-C completion handler -> async bridging. I predict we see an **unprecedented** number of Apple API deprecations in the coming years because of this. Same thing is going to happen to app internals. It is basically impossible to use concurrency without running into issues. I believe this is good. Developing the skills to deal with these problems and getting **comfortable** with constructs like `nonisolated(unsafe)` should be a high priority for all Swift developers right now.

That my take. I hope you are right that complete checking is becoming fashionable!

**Reply by seanmrich** [reply]
> Mark let globals as `nonisolated(unsafe)`

I’m glad you said this, because it gives me a chance to ask directly. I’m nervous about using `nonisolated(unsafe)` for two reasons:
- With a library as complex as GRDB it’s difficult to be sure that I understand the nuances well enough to be confident that there aren’t concurrency issues, now or in future revisions. Should I feel confident in using things like static associations or `Column` definitions with `nonisolated(unsafe)`?
- Once sendability is supported, there’s no indication by the compiler that I should remove the annotation. Diligent reading of release notes is the only trigger. Am I wrong?

**Reply by groue** [reply]
@seanmrich, you're correct to feel nervous about `nonisolated(unsafe)`. But I will also try to soothe your concerns.

First, I really want to ship a GRDB version where associations are Sendable, and your original code compiles without any warning:

```swift
// GRDB 7
extension Author {
  // 😇 No warning because HasManyAssociation is Sendable
  static let books = hasMany(Book.self)
}
```

Meanwhile, a way to remove the warning is `nonisolated(unsafe)`:

```swift
// GRDB 6
extension Author {
  // 😰
  nonisolated(unsafe) static let books = hasMany(Book.self)
}
```

Now let's soothe your concerns:

The `hasMany(Book.self)` association is completely safe. You will not have any problem at runtime. The _type_ of this value is not marked as `Sendable`, which explains why the compiler warns. But this _particular instance_ of association has all the runtime properties of `Sendable` values. It won't blow up your program, 100% guaranteed. But you have to trust my words, because the compiler warns.

And the compiler warns for a good reason: the current public apis of GRDB _makes it possible_ to build non-sendable and unsafe associations. On the other hand, your app much probably does not.

For example:

```swift
extension Author {
  // 👎 A DANGEROUS ASSOCIATION
  static func wackyBooks(_ value: NonSendableType) -> HasManyAssociation<Author, Book> {
    Author.hasMany(Book.self).filter(Column("whatever") == value)
  }
}

let value: NonSendableType = ...
let authors = try await dbQueue.read { db in
  try Author
    .having(Author.wackyBooks(value).isEmpty) // 😅
    .fetchAll(db)
}
```

See? It is _possible_ to define associations that involve non-sendable values, and write code that has this non-sendable value cross isolation domains in an unsafe way.

This _possibility_ is the reason why GRDB associations are not Sendable today, and the reason why you need to use `nonisolated(unsafe)` in order to mute the current warning.

Fortunately, most requests and associations are built from plain strings, numbers, etc. You can audit your code for requests and associations built out of non-sendable values, but chances are low. And I'd bet that no _association_ uses such values.

I hope I could convince you that decorating plain associatons with `nonisolated(unsafe)` is ugly, but does not create any risk.

In GRDB 7, the library will _require_ that requests and associations are built from `Sendable` values. The compiler will warn on the definition of `wackyBooks`, above. But it will no longer warn on `static let books = hasMany(Book.self)`.

---

EDIT: the `wackyBooks` association actually is safe, because it does not embed the non-sendable value (it embeds a database representation of the value, which is safe). The problem with the above sample code is not the association, but the fact that `read() async throws` does not require a `@Sendable` closure (it will in GRDB 7).

A really unsafe association would _embed_ the non-sendable value inside it, for example with a closure:

```swift
extension Author {
  // 👿 DEMONIAC ASSOCIATION
  static func wackyBooks(_ value: NonSendableType) -> HasManyAssociation<Author, Book> {
    Author.hasMany(Book.self).filterWhenConnected { db in
      Column("whatever") == value
    }
  }
}
```

Now this one is nasty, especially if you use such an association with tools like database observation that will run the request at about any time from various database threads - typically a context where we really need `Sendable` :-)

_Fortunately, nobody writes code like that._

**Reply by mattmassicotte** [reply]
> Once sendability is supported, there’s no indication by the compiler that I should remove the annotation. Diligent reading of release notes is the only trigger. Am I wrong?

You are not wrong. But, the compiler team seems [interested](https://github.com/apple/swift/issues/71546) in catching this. However, I do want to point out that this is purely a cosmetic problem. If this is used to express safety that is just invisible to the compiler, like @groue explained in fantastic detail, nothing changes.

**Reply by groue** [reply]
> Developing the skills to deal with these problems and getting **comfortable** with constructs like `nonisolated(unsafe)` should be a high priority for all Swift developers right now.

You're right, @mattmassicotte. I also sometimes feel like we're living in a bubble, surrounded by the most demanding developers who won't rest until they have a clear understanding of Swift concurrency and related techniques. This explains all our questions, frustrations, incessant explorations, speeches, writings, conversations with our peers... I sometimes wonder if this level of demand, impatience and rigor is _normal_ and should be expected from all devs. 😅

**Reply by seanmrich** [reply]
This was all tremendously helpful for me. Concurrency is notoriously challenging. I’ve definitely screwed it up in the past so I tend to wring my hands a bit, especially since the idea of the compiler preventing data races (trying at least) is new to me. I truly appreciate the guidance.

**Reply by mattmassicotte** [reply]
> I sometimes wonder if this level of demand, impatience and rigor is normal and should be expected from all devs.

I certainly hope not! But, changes like this don't happen very often.

The warnings are opt-in for a reason. There are huge trade-offs, but one advantage is avoiding confusion/frustration until Swift and Apple's own APIs are ready and they have some documentation on how everything is supposed to work. We're beta testers!
